actors/ppo.py

```diff
@@ -0,0 +1,720 @@
+
+                    # Publish the newly updated model
+                    self.model.iter_idx = iter_idx
+                    version_info = await run_session.model_registry.publish_version(self.model)
+            print(f"Iter #{iter_idx}: step: {len(actions)}")
+
+    async def train_step(
+        self,
+        observations: List[torch.Tensor],
+        rewards: List[torch.Tensor],
+        actions: List[torch.Tensor],
+        dones: List[torch.Tensor],
+    ):
+        """Train the model after collecting the data from the trial"""
+
+        # Take n steps from the rollout
+        observations = torch.vstack(observations)[: self._cfg.num_steps * self._cfg.epoch_num_trials + 1]
+        actions = torch.vstack(actions)[: self._cfg.num_steps * self._cfg.epoch_num_trials]
+        rewards = torch.vstack(rewards)[: self._cfg.num_steps * self._cfg.epoch_num_trials]
+        dones = torch.vstack(dones)[: self._cfg.num_steps * self._cfg.epoch_num_trials]
+
+        # Normalize the observations
+        if self.model.state_normalization is not None:
+            self.model.state_normalization.update(observations)
+
+        # Make a dataloader in order to process data in batch
+        batch_state = self.make_dataloader(observations[:-1], self._cfg.batch_size, self.model._num_input)
+        batch_action = self.make_dataloader(actions, self._cfg.batch_size, self.model._num_output)
+
+        values = self.compute_value(batch_state)
+        with torch.no_grad():
+            next_value = self.model.value_network(observations[-1:])
+        next_value = next_value * (1 - dones[-1])
+        values = torch.cat((values, next_value), dim=0)
+
+        log_probs = self.compute_log_lik(batch_state, batch_action)
+
+        # Compute the generalized advantage estimation
+        advs = self.compute_gae(
+            rewards=rewards, values=values, dones=dones, gamma=self._cfg.discount_factor, lam=self._cfg.lambda_gae
+        )
+
+        # Update parameters for policy and value networks
+        policy_loss, value_loss = self.update_parameters(
+            states=observations[:-1],
+            actions=actions,
+            advs=advs,
+            values=values[:-1],
+            log_probs=log_probs,
+            num_epochs=self._cfg.num_epochs,
+        )
+
+        return policy_loss, value_loss
+
+    def update_parameters(
+        self,
+        states: torch.Tensor,
+        actions: torch.Tensor,
+        advs: torch.Tensor,
+        values: torch.Tensor,
+        log_probs: torch.Tensor,
+        num_epochs: int,
+    ) -> Tuple[torch.Tensor]:
+        """Update policy & value networks"""
+
+        returns = advs + values
+        num_states = len(returns)
+        for _ in range(num_epochs):
+            for _ in range(num_states // self._cfg.batch_size):
+                # Get data in batch. TODO: Send data to device (need to test with cuda)
+                idx = np.random.randint(0, num_states, self._cfg.batch_size)
```

Comment:
You may create a rng = np.random.default_rng() outside the loop and use it here

Reply:
I think ` np.random.default_rng()` for a wide range of distributions while here I only target the discrete uniform for the interger only